src/components/ProductFilter.js

```diff
@@ -0,0 +1,278 @@
+import { useEffect, useMemo, useReducer } from "react";
+import { useState } from "react";
+import "./css/ProductFilter.css";
+
+const ProductFilter = ({ setFilter, className, buttonFun, data }) => {
+  //ready to get irritated as i got seeing my own code why wont you
+
+  const [state, dispatch] = useReducer(FilterReducer, {
```

Comment:
You can destructure this state so that you don't have to write `state.` in multiple places